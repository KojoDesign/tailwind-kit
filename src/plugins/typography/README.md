# Typography

It's really unfortunate Tailwind has an official [plugin of the same name](https://tailwindcss.com/docs/typography-plugin). It's kind of silly, considering it's a plugin called `typography`, but all the class names its generates are prefixed with `prose-`.

The TailwindKit Typography plugin is a tad more conventional in the sense that it helps you build your own typographic systems like those found in design systems such as [Material Design](https://m3.material.io/styles/typography/type-scale-tokens).

## Features

### Custom Typography

The main purpose of the typography plugin is to create your own typographic systems. Using the plugin, you can easily group font families, styles, and weights to create consistent typography across your UI.

In your theme under the `typography` key (don't worry â€“ this can be configured to avoid conflicts) you can easily specify variants:

```javascript
module.exports = {
  theme: {
    typography: ({ theme }: any) => ({
      variants: {
        headline: {
          family: theme('fontFamily.display'),
          leading: theme('lineHeight.tight'),
          tracking: theme('letterSpacing.tight')
        },
        DEFAULT: {
          family: theme('fontFamily.sans'),
          leading: theme('lineHeight.tight'),
        }
      }
    })
  }
}
```

This config will generate the classes `typography` (everything in `DEFAULT`) and `typography-headline`. Similar to a component library, you'll need to use both classes for a headline element to have all CSS work correctly:

```html
<h1 class="typography typography-headline">
  Hello world
</h1>
```

Notice how the config doesn't use freeform CSS-in-JS. The plugin needs to perform some additional work under-the-hood with the values provided, so the config only accepts certain keys.

They are:

- **family:** Font family
- **leading:** Line height
- **tracking:** Letter spacing
- **weight:** Font weight
- **size:** Font size

### Dynamic Sizes

The Typography plugin generates a base `typography` class, as well as variants such as `typography-headline`, with which you can style text to have a specific font family, weight and style. Think of it as a grouping of `font-` and `text-` classes.

You can also add size variants, such as `typography-sm`, that can apply different sizes based on the variant being styled. This follows Material Design's pattern of having [different sizes](https://m3.material.io/styles/typography/type-scale-tokens#40fc37f8-3269-4aa3-9f28-c6113079ac5d) for each main typographic variant.

You can also pair this plugin with the [typescale plugin](../typescale/README.md) to generate a consistent type scale to use in your typography.

## Configuring the Plugin

From your Tailwind theme you can easily add typography and size variants:

```javascript
module.exports = {
  // ...
  theme: {
    // ...
    typography: ({ theme }: any) => ({
      colors: {
        muted: theme('colors.slate.900/0.55'),
        primary: theme('colors.purple.400'),
        subtle: theme('colors.slate.900/0.8'),
        default: {
          light: theme('colors.slate.900'),
          dark: theme('colors.white'),
        },
      },
      sizes: ['sm', 'lg'],
      variants: {
        DEFAULT: {
          family: theme('fontFamily.sans'),
          leading: theme('lineHeight.tight'),
        },
        display: {
          weight: theme('fontWeight.bold'),
          leading: theme('lineHeight.none'),
          size: theme('fontSize.5xl'),
        },
        headline: {
          weight: theme('fontWeight.bold'),
          tracking: theme('letterSpacing.tight'),
          size: theme('fontSize.4xl'),
        }
      }
    })
  }
}
```