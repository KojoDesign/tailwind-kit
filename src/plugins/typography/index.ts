export * from "./typography.plugin";
export * from "./typography.types";
