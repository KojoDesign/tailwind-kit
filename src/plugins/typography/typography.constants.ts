import { GLOBAL_VARIABLE_PREFIX } from "../../constants";
import { createVariableName } from "../utils";

export const VARIABLE_PREFIX = createVariableName(
  GLOBAL_VARIABLE_PREFIX,
  "typography",
);

export const VARIABLE_COLOR_BASE = createVariableName(VARIABLE_PREFIX, "color");

export const VARIABLE_GRADIENT_STOPS = createVariableName(
  VARIABLE_PREFIX,
  "gradient-stops",
);

export const VARIABLE_COLOR_INVERTED = createVariableName(
  VARIABLE_PREFIX,
  "color-inverted",
);

export const VARIABLE_LEADING = createVariableName(VARIABLE_PREFIX, "leading");

export const VARIABLE_FONT_SIZE = createVariableName(
  VARIABLE_PREFIX,
  "font-size",
);
