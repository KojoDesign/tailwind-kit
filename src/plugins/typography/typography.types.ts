export type TypographyProperty =
  | "direction"
  | "fontFamily"
  | "fontFeatureSettings"
  | "fontKerning"
  | "fontOpticalSizing"
  | "fontStretch"
  | "fontStyle"
  | "fontVariantAlternates"
  | "fontVariantCaps"
  | "fontVariantEastAsian"
  | "fontVariantLigatures"
  | "fontVariantNumeric"
  | "fontVariantPosition"
  | "fontVariationSettings"
  | "fontWeight"
  | "letterSpacing"
  | "lineHeight"
  | "tabSize"
  | "textAlign"
  | "textAlignLast"
  | "textDecorationColor"
  | "textDecorationLine"
  | "textDecorationStyle"
  | "textDecorationThickness"
  | "textIndent"
  | "textOverflow"
  | "textShadow"
  | "textSizeAdjust"
  | "textTransform"
  | "textWrap"
  | "verticalAlign"
  | "whiteSpaceCollapse"
  | "wordBreak"
  | "wordSpacing";

export type TypographyCSS = Partial<Record<TypographyProperty, string>>;

export type ColorValue = string | { light: string; dark: string };

export type FontSizes<T> = Record<string, ((variant: T) => string) | string>;

export type TypographyTheme<T extends string = string> = {
  base: TypographyCSS;
  variants: Record<T, TypographyCSS>;
  sizes: FontSizes<T>;
  colors: Record<string, ColorValue>;
};

export interface TypographyOptions {
  classPrefix: string;
  themeKey: string;
  gradient: {
    toOpacity: number;
    multiline: boolean;
  };
}
