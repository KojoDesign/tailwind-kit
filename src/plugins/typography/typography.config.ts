import { ResolvableTo } from "tailwindcss/types/config";

import { TypographyOptions, TypographyTheme } from "./typography.types";

export const defaultTheme: ResolvableTo<TypographyTheme> = ({ theme }) => ({
  base: {},
  sizes: {},
  variants: {},
  colors: {
    DEFAULT: theme("colors.gray.700"),
  },
});

export const defaultOptions: TypographyOptions = {
  classPrefix: "typography",
  themeKey: "typography",
  gradient: {
    toOpacity: 0.8,
    multiline: true,
  },
};
