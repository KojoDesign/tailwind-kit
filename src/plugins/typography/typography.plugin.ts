import kebabCase from "kebab-case";
import plugin from "tailwindcss/plugin";

import { LINEAR_DIRECTIONS } from "../constants";
import {
  createClassName,
  createVariableName,
  referenceVariable,
} from "../utils";
import { defaultOptions, defaultTheme } from "./typography.config";
import {
  VARIABLE_COLOR_BASE,
  VARIABLE_COLOR_INVERTED,
  VARIABLE_FONT_SIZE,
  VARIABLE_GRADIENT_STOPS,
  VARIABLE_LEADING,
} from "./typography.constants";
import { TypographyOptions, TypographyTheme } from "./typography.types";
import { getColorObject, getFullOptions } from "./typography.utils";

const { withOptions } = plugin;

const typography = withOptions<Partial<TypographyOptions>>(
  (options) =>
    ({ addUtilities, matchUtilities, addComponents, e, theme }) => {
      const { classPrefix, gradient, themeKey } = getFullOptions(options);

      const { base, colors, sizes, variants } = theme(
        themeKey,
      ) as TypographyTheme;

      const { light: defaultLight, dark: defaultDark } = getColorObject(
        colors?.["DEFAULT"] ?? "inherit",
      );

      const defaultFontSize =
        (typeof sizes?.["DEFAULT"] === "function"
          ? sizes?.["DEFAULT"]("")
          : sizes?.["DEFAULT"]) ?? "inherit";

      const baseLineHeight = base?.lineHeight?.toString() ?? "1em";

      const defaultGradientStops = [
        `rgba(255, 255, 255, ${gradient.toOpacity})`,
        `rgba(255, 255, 255, 1)`,
      ];

      // Add color variants
      addUtilities(
        Object.entries(colors ?? {}).map(([name, color]) => {
          const { light, dark } = getColorObject(color);
          return {
            [createClassName(classPrefix, e(name))]: {
              [VARIABLE_COLOR_BASE]: light,
              [VARIABLE_COLOR_INVERTED]: dark,
              color: referenceVariable(VARIABLE_COLOR_BASE),
            },
          };
        }),
      );

      // Add size utilities
      addUtilities(
        Object.keys(sizes ?? {}).map((name) => ({
          [createClassName(classPrefix, e(name))]: {
            [VARIABLE_FONT_SIZE]: referenceVariable(
              createVariableName(VARIABLE_FONT_SIZE, e(name)),
            ),
          },
        })),
      );

      // Add base typography component
      addComponents({
        [`.${classPrefix}`]: {
          ...base,
          [VARIABLE_COLOR_BASE]: defaultLight,
          [VARIABLE_COLOR_INVERTED]: defaultDark,
          [VARIABLE_FONT_SIZE]: defaultFontSize,
          [VARIABLE_LEADING]: baseLineHeight,
          color: referenceVariable(VARIABLE_COLOR_BASE),
          fontSize: referenceVariable(VARIABLE_FONT_SIZE),
          fontFamily:
            (Array.isArray(base?.fontFamily)
              ? base?.fontFamily.join(", ")
              : base?.fontFamily) ?? "inherit",
        },
      });

      // Add typographic variants
      addComponents(
        Object.entries(variants ?? {}).map(([name, rules]) => {
          let sizeVariables = {};

          sizeVariables = Object.entries(sizes ?? {}).reduce(
            (acc, [sizeName, size]) => {
              const customSize = typeof size === "function" ? size(name) : size;

              const variableName =
                sizeName === "DEFAULT"
                  ? VARIABLE_FONT_SIZE
                  : createVariableName(VARIABLE_FONT_SIZE, e(sizeName));

              return {
                ...acc,
                ...(customSize != null && { [variableName]: customSize }),
              };
            },
            {},
          );

          return {
            [createClassName(classPrefix, e(name))]: {
              ...rules,
              ...sizeVariables,
              [VARIABLE_LEADING]: rules.lineHeight ?? "1em",
              fontFamily:
                (Array.isArray(rules?.fontFamily)
                  ? rules?.fontFamily.join(", ")
                  : rules?.fontFamily) ?? "inherit",
            },
          };
        }),
      );

      matchUtilities(
        {
          [createClassName(classPrefix, "gradient-to").substring(1)]: (
            value,
          ) => ({
            [VARIABLE_GRADIENT_STOPS]: defaultGradientStops
              .map((s, i) => `${s} ${i * 100}%`)
              .join(", "),
            maskImage: `linear-gradient(${value}, ${referenceVariable(
              VARIABLE_GRADIENT_STOPS,
            )})`,
            maskRepeat: "repeat-y",
            maskSize: `100% calc(${referenceVariable(VARIABLE_LEADING)} * 1em)`,
          }),
        },
        {
          values: theme("linearDirections"),
        },
      );

      // Add invert utility
      addUtilities({
        [createClassName(classPrefix, "invert")]: {
          [VARIABLE_COLOR_BASE]: referenceVariable(VARIABLE_COLOR_INVERTED),
          [VARIABLE_GRADIENT_STOPS]: defaultGradientStops
            .slice()
            .reverse()
            .map((s, i) => `${s} ${i * 100}%`)
            .join(", "),
        },
      });
    },
  ({ themeKey = defaultOptions.themeKey } = defaultOptions) => ({
    theme: {
      [themeKey]: defaultTheme,
      linearDirections: LINEAR_DIRECTIONS,
    },
  }),
);

export { typography, defaultTheme };
