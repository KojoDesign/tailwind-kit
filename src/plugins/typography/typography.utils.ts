import { TinyColor } from "@ctrl/tinycolor";

import merge from "deepmerge";

import { defaultOptions } from "./typography.config";
import { ColorValue, TypographyOptions } from "./typography.types";

export function invertColor(color: string) {
  const tc = new TinyColor(color);
  const hsv = tc.toHsv();
  const complement = 1 - hsv.v;

  return new TinyColor({
    h: hsv.h,
    s: 0,
    v: complement * 100,
  }).toRgbString();
}

export function getColorObject(color: ColorValue) {
  let light = color as string;
  let dark = null;

  if (typeof color === "object") {
    light = color.light;
    dark = color.dark;
  }

  if (!dark) {
    dark = invertColor(light);
  }

  return { light, dark };
}

export function getFullOptions(
  options?: Partial<TypographyOptions>,
): TypographyOptions {
  return merge(defaultOptions, options ?? {});
}
