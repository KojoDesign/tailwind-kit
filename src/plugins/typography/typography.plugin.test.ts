import "../../testing/matchers";

import { generateCSS, html } from "../../testing/index";
import { typography } from "./typography.plugin";

test("generates variants correctly", async () => {
  const css = await generateCSS(
    html`<div class="typography typography-headline"></div>`,
    {
      plugins: [typography],
      theme: {
        typography: {
          variants: {
            DEFAULT: {
              leading: "1",
            },
            headline: {
              family: "red",
            },
          },
        },
      },
    },
  );

  expect(css).toMatchSnapshot();
});

test("generates colors correctly", async () => {
  const css = await generateCSS(
    html`<div
      class="typography typography-primary typography-subtle typography-invert"
    ></div>`,
    {
      plugins: [typography],
      theme: {
        typography: {
          colors: {
            subtle: "rgb(40, 40, 40)",
            primary: {
              light: "red",
              dark: "green",
            },
          },
        },
      },
    },
  );

  expect(css).toMatchSnapshot();
});

test("generates sizes correctly", async () => {
  const css = await generateCSS(
    html`<div class="typography typography-headline typography-sm"></div>`,
    {
      plugins: [typography],
      theme: {
        typography: {
          variants: {
            headline: {
              family: "Lato",
            },
          },
          sizes: {
            sm: (variant: string) => {
              if (variant === "headline") {
                return "2rem";
              }
              return "1rem";
            },
          },
        },
      },
    },
  );

  expect(css).toMatchSnapshot();
});

test("generates gradients correctly", async () => {
  const css = await generateCSS(
    html`<div
      class="typography typography-headline typography-gradient-to-b typography-sm"
    ></div>`,
    {
      plugins: [typography],
      theme: {
        typography: {
          variants: {
            headline: {
              family: "Lato",
            },
          },
          sizes: {
            sm: (variant: string) => {
              if (variant === "headline") {
                return "2rem";
              }
              return "1rem";
            },
          },
        },
      },
    },
  );

  expect(css).toMatchSnapshot();
});
