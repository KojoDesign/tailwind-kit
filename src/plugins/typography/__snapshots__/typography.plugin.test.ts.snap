// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates colors correctly 1`] = `
"*, ::before, ::after {
    --tw-border-spacing-x: 0;
    --tw-border-spacing-y: 0;
    --tw-translate-x: 0;
    --tw-translate-y: 0;
    --tw-rotate: 0;
    --tw-skew-x: 0;
    --tw-skew-y: 0;
    --tw-scale-x: 1;
    --tw-scale-y: 1;
    --tw-pan-x:  ;
    --tw-pan-y:  ;
    --tw-pinch-zoom:  ;
    --tw-scroll-snap-strictness: proximity;
    --tw-gradient-from-position:  ;
    --tw-gradient-via-position:  ;
    --tw-gradient-to-position:  ;
    --tw-ordinal:  ;
    --tw-slashed-zero:  ;
    --tw-numeric-figure:  ;
    --tw-numeric-spacing:  ;
    --tw-numeric-fraction:  ;
    --tw-ring-inset:  ;
    --tw-ring-offset-width: 0px;
    --tw-ring-offset-color: #fff;
    --tw-ring-color: rgb(59 130 246 / 0.5);
    --tw-ring-offset-shadow: 0 0 #0000;
    --tw-ring-shadow: 0 0 #0000;
    --tw-shadow: 0 0 #0000;
    --tw-shadow-colored: 0 0 #0000;
    --tw-blur:  ;
    --tw-brightness:  ;
    --tw-contrast:  ;
    --tw-grayscale:  ;
    --tw-hue-rotate:  ;
    --tw-invert:  ;
    --tw-saturate:  ;
    --tw-sepia:  ;
    --tw-drop-shadow:  ;
    --tw-backdrop-blur:  ;
    --tw-backdrop-brightness:  ;
    --tw-backdrop-contrast:  ;
    --tw-backdrop-grayscale:  ;
    --tw-backdrop-hue-rotate:  ;
    --tw-backdrop-invert:  ;
    --tw-backdrop-opacity:  ;
    --tw-backdrop-saturate:  ;
    --tw-backdrop-sepia:  
}
::backdrop {
    --tw-border-spacing-x: 0;
    --tw-border-spacing-y: 0;
    --tw-translate-x: 0;
    --tw-translate-y: 0;
    --tw-rotate: 0;
    --tw-skew-x: 0;
    --tw-skew-y: 0;
    --tw-scale-x: 1;
    --tw-scale-y: 1;
    --tw-pan-x:  ;
    --tw-pan-y:  ;
    --tw-pinch-zoom:  ;
    --tw-scroll-snap-strictness: proximity;
    --tw-gradient-from-position:  ;
    --tw-gradient-via-position:  ;
    --tw-gradient-to-position:  ;
    --tw-ordinal:  ;
    --tw-slashed-zero:  ;
    --tw-numeric-figure:  ;
    --tw-numeric-spacing:  ;
    --tw-numeric-fraction:  ;
    --tw-ring-inset:  ;
    --tw-ring-offset-width: 0px;
    --tw-ring-offset-color: #fff;
    --tw-ring-color: rgb(59 130 246 / 0.5);
    --tw-ring-offset-shadow: 0 0 #0000;
    --tw-ring-shadow: 0 0 #0000;
    --tw-shadow: 0 0 #0000;
    --tw-shadow-colored: 0 0 #0000;
    --tw-blur:  ;
    --tw-brightness:  ;
    --tw-contrast:  ;
    --tw-grayscale:  ;
    --tw-hue-rotate:  ;
    --tw-invert:  ;
    --tw-saturate:  ;
    --tw-sepia:  ;
    --tw-drop-shadow:  ;
    --tw-backdrop-blur:  ;
    --tw-backdrop-brightness:  ;
    --tw-backdrop-contrast:  ;
    --tw-backdrop-grayscale:  ;
    --tw-backdrop-hue-rotate:  ;
    --tw-backdrop-invert:  ;
    --tw-backdrop-opacity:  ;
    --tw-backdrop-saturate:  ;
    --tw-backdrop-sepia:  
}
.typography {
    --kit-typography-color: inherit;
    --kit-typography-color-inverted: rgb(255, 255, 255);
    --kit-typography-font-size: inherit;
    --kit-typography-leading: 1em;
    color: var(--kit-typography-color);
    font-size: var(--kit-typography-font-size);
    font-family: inherit
}
.typography-subtle {
    --kit-typography-color: rgb(40, 40, 40);
    --kit-typography-color-inverted: rgb(215, 215, 215)
}
.typography-primary {
    --kit-typography-color: red;
    --kit-typography-color-inverted: green
}
.typography-invert {
    --kit-typography-color: var(--kit-typography-color-inverted);
    --kit-typography-gradient-stops: rgba(255, 255, 255, 1) 100%, rgba(255, 255, 255, 0.8) 0%
}"
`;

exports[`generates gradients correctly 1`] = `
"*, ::before, ::after {
    --tw-border-spacing-x: 0;
    --tw-border-spacing-y: 0;
    --tw-translate-x: 0;
    --tw-translate-y: 0;
    --tw-rotate: 0;
    --tw-skew-x: 0;
    --tw-skew-y: 0;
    --tw-scale-x: 1;
    --tw-scale-y: 1;
    --tw-pan-x:  ;
    --tw-pan-y:  ;
    --tw-pinch-zoom:  ;
    --tw-scroll-snap-strictness: proximity;
    --tw-gradient-from-position:  ;
    --tw-gradient-via-position:  ;
    --tw-gradient-to-position:  ;
    --tw-ordinal:  ;
    --tw-slashed-zero:  ;
    --tw-numeric-figure:  ;
    --tw-numeric-spacing:  ;
    --tw-numeric-fraction:  ;
    --tw-ring-inset:  ;
    --tw-ring-offset-width: 0px;
    --tw-ring-offset-color: #fff;
    --tw-ring-color: rgb(59 130 246 / 0.5);
    --tw-ring-offset-shadow: 0 0 #0000;
    --tw-ring-shadow: 0 0 #0000;
    --tw-shadow: 0 0 #0000;
    --tw-shadow-colored: 0 0 #0000;
    --tw-blur:  ;
    --tw-brightness:  ;
    --tw-contrast:  ;
    --tw-grayscale:  ;
    --tw-hue-rotate:  ;
    --tw-invert:  ;
    --tw-saturate:  ;
    --tw-sepia:  ;
    --tw-drop-shadow:  ;
    --tw-backdrop-blur:  ;
    --tw-backdrop-brightness:  ;
    --tw-backdrop-contrast:  ;
    --tw-backdrop-grayscale:  ;
    --tw-backdrop-hue-rotate:  ;
    --tw-backdrop-invert:  ;
    --tw-backdrop-opacity:  ;
    --tw-backdrop-saturate:  ;
    --tw-backdrop-sepia:  
}
::backdrop {
    --tw-border-spacing-x: 0;
    --tw-border-spacing-y: 0;
    --tw-translate-x: 0;
    --tw-translate-y: 0;
    --tw-rotate: 0;
    --tw-skew-x: 0;
    --tw-skew-y: 0;
    --tw-scale-x: 1;
    --tw-scale-y: 1;
    --tw-pan-x:  ;
    --tw-pan-y:  ;
    --tw-pinch-zoom:  ;
    --tw-scroll-snap-strictness: proximity;
    --tw-gradient-from-position:  ;
    --tw-gradient-via-position:  ;
    --tw-gradient-to-position:  ;
    --tw-ordinal:  ;
    --tw-slashed-zero:  ;
    --tw-numeric-figure:  ;
    --tw-numeric-spacing:  ;
    --tw-numeric-fraction:  ;
    --tw-ring-inset:  ;
    --tw-ring-offset-width: 0px;
    --tw-ring-offset-color: #fff;
    --tw-ring-color: rgb(59 130 246 / 0.5);
    --tw-ring-offset-shadow: 0 0 #0000;
    --tw-ring-shadow: 0 0 #0000;
    --tw-shadow: 0 0 #0000;
    --tw-shadow-colored: 0 0 #0000;
    --tw-blur:  ;
    --tw-brightness:  ;
    --tw-contrast:  ;
    --tw-grayscale:  ;
    --tw-hue-rotate:  ;
    --tw-invert:  ;
    --tw-saturate:  ;
    --tw-sepia:  ;
    --tw-drop-shadow:  ;
    --tw-backdrop-blur:  ;
    --tw-backdrop-brightness:  ;
    --tw-backdrop-contrast:  ;
    --tw-backdrop-grayscale:  ;
    --tw-backdrop-hue-rotate:  ;
    --tw-backdrop-invert:  ;
    --tw-backdrop-opacity:  ;
    --tw-backdrop-saturate:  ;
    --tw-backdrop-sepia:  
}
.typography {
    --kit-typography-color: inherit;
    --kit-typography-color-inverted: rgb(255, 255, 255);
    --kit-typography-font-size: inherit;
    --kit-typography-leading: 1em;
    color: var(--kit-typography-color);
    font-size: var(--kit-typography-font-size);
    font-family: inherit
}
.typography-headline {
    font-family: Lato;
    --kit-typography-font-size-sm: 2rem;
    --kit-typography-leading: 1em
}
.typography-sm {
    --kit-typography-font-size: var(--kit-typography-font-size-sm)
}
.typography-gradient-to-b {
    --kit-typography-gradient-stops: rgba(255, 255, 255, 1) 100%, rgba(255, 255, 255, 0.8) 0%;
    mask-image: linear-gradient(to bottom, var(--kit-typography-gradient-stops));
    mask-repeat: repeat-y;
    mask-size: 100% calc(var(--kit-typography-leading) * 1em)
}"
`;

exports[`generates sizes correctly 1`] = `
"*, ::before, ::after {
    --tw-border-spacing-x: 0;
    --tw-border-spacing-y: 0;
    --tw-translate-x: 0;
    --tw-translate-y: 0;
    --tw-rotate: 0;
    --tw-skew-x: 0;
    --tw-skew-y: 0;
    --tw-scale-x: 1;
    --tw-scale-y: 1;
    --tw-pan-x:  ;
    --tw-pan-y:  ;
    --tw-pinch-zoom:  ;
    --tw-scroll-snap-strictness: proximity;
    --tw-gradient-from-position:  ;
    --tw-gradient-via-position:  ;
    --tw-gradient-to-position:  ;
    --tw-ordinal:  ;
    --tw-slashed-zero:  ;
    --tw-numeric-figure:  ;
    --tw-numeric-spacing:  ;
    --tw-numeric-fraction:  ;
    --tw-ring-inset:  ;
    --tw-ring-offset-width: 0px;
    --tw-ring-offset-color: #fff;
    --tw-ring-color: rgb(59 130 246 / 0.5);
    --tw-ring-offset-shadow: 0 0 #0000;
    --tw-ring-shadow: 0 0 #0000;
    --tw-shadow: 0 0 #0000;
    --tw-shadow-colored: 0 0 #0000;
    --tw-blur:  ;
    --tw-brightness:  ;
    --tw-contrast:  ;
    --tw-grayscale:  ;
    --tw-hue-rotate:  ;
    --tw-invert:  ;
    --tw-saturate:  ;
    --tw-sepia:  ;
    --tw-drop-shadow:  ;
    --tw-backdrop-blur:  ;
    --tw-backdrop-brightness:  ;
    --tw-backdrop-contrast:  ;
    --tw-backdrop-grayscale:  ;
    --tw-backdrop-hue-rotate:  ;
    --tw-backdrop-invert:  ;
    --tw-backdrop-opacity:  ;
    --tw-backdrop-saturate:  ;
    --tw-backdrop-sepia:  
}
::backdrop {
    --tw-border-spacing-x: 0;
    --tw-border-spacing-y: 0;
    --tw-translate-x: 0;
    --tw-translate-y: 0;
    --tw-rotate: 0;
    --tw-skew-x: 0;
    --tw-skew-y: 0;
    --tw-scale-x: 1;
    --tw-scale-y: 1;
    --tw-pan-x:  ;
    --tw-pan-y:  ;
    --tw-pinch-zoom:  ;
    --tw-scroll-snap-strictness: proximity;
    --tw-gradient-from-position:  ;
    --tw-gradient-via-position:  ;
    --tw-gradient-to-position:  ;
    --tw-ordinal:  ;
    --tw-slashed-zero:  ;
    --tw-numeric-figure:  ;
    --tw-numeric-spacing:  ;
    --tw-numeric-fraction:  ;
    --tw-ring-inset:  ;
    --tw-ring-offset-width: 0px;
    --tw-ring-offset-color: #fff;
    --tw-ring-color: rgb(59 130 246 / 0.5);
    --tw-ring-offset-shadow: 0 0 #0000;
    --tw-ring-shadow: 0 0 #0000;
    --tw-shadow: 0 0 #0000;
    --tw-shadow-colored: 0 0 #0000;
    --tw-blur:  ;
    --tw-brightness:  ;
    --tw-contrast:  ;
    --tw-grayscale:  ;
    --tw-hue-rotate:  ;
    --tw-invert:  ;
    --tw-saturate:  ;
    --tw-sepia:  ;
    --tw-drop-shadow:  ;
    --tw-backdrop-blur:  ;
    --tw-backdrop-brightness:  ;
    --tw-backdrop-contrast:  ;
    --tw-backdrop-grayscale:  ;
    --tw-backdrop-hue-rotate:  ;
    --tw-backdrop-invert:  ;
    --tw-backdrop-opacity:  ;
    --tw-backdrop-saturate:  ;
    --tw-backdrop-sepia:  
}
.typography {
    --kit-typography-color: inherit;
    --kit-typography-color-inverted: rgb(255, 255, 255);
    --kit-typography-font-size: inherit;
    --kit-typography-leading: 1em;
    color: var(--kit-typography-color);
    font-size: var(--kit-typography-font-size);
    font-family: inherit
}
.typography-headline {
    font-family: Lato;
    --kit-typography-font-size-sm: 2rem;
    --kit-typography-leading: 1em
}
.typography-sm {
    --kit-typography-font-size: var(--kit-typography-font-size-sm)
}"
`;

exports[`generates variants correctly 1`] = `
"*, ::before, ::after {
    --tw-border-spacing-x: 0;
    --tw-border-spacing-y: 0;
    --tw-translate-x: 0;
    --tw-translate-y: 0;
    --tw-rotate: 0;
    --tw-skew-x: 0;
    --tw-skew-y: 0;
    --tw-scale-x: 1;
    --tw-scale-y: 1;
    --tw-pan-x:  ;
    --tw-pan-y:  ;
    --tw-pinch-zoom:  ;
    --tw-scroll-snap-strictness: proximity;
    --tw-gradient-from-position:  ;
    --tw-gradient-via-position:  ;
    --tw-gradient-to-position:  ;
    --tw-ordinal:  ;
    --tw-slashed-zero:  ;
    --tw-numeric-figure:  ;
    --tw-numeric-spacing:  ;
    --tw-numeric-fraction:  ;
    --tw-ring-inset:  ;
    --tw-ring-offset-width: 0px;
    --tw-ring-offset-color: #fff;
    --tw-ring-color: rgb(59 130 246 / 0.5);
    --tw-ring-offset-shadow: 0 0 #0000;
    --tw-ring-shadow: 0 0 #0000;
    --tw-shadow: 0 0 #0000;
    --tw-shadow-colored: 0 0 #0000;
    --tw-blur:  ;
    --tw-brightness:  ;
    --tw-contrast:  ;
    --tw-grayscale:  ;
    --tw-hue-rotate:  ;
    --tw-invert:  ;
    --tw-saturate:  ;
    --tw-sepia:  ;
    --tw-drop-shadow:  ;
    --tw-backdrop-blur:  ;
    --tw-backdrop-brightness:  ;
    --tw-backdrop-contrast:  ;
    --tw-backdrop-grayscale:  ;
    --tw-backdrop-hue-rotate:  ;
    --tw-backdrop-invert:  ;
    --tw-backdrop-opacity:  ;
    --tw-backdrop-saturate:  ;
    --tw-backdrop-sepia:  
}
::backdrop {
    --tw-border-spacing-x: 0;
    --tw-border-spacing-y: 0;
    --tw-translate-x: 0;
    --tw-translate-y: 0;
    --tw-rotate: 0;
    --tw-skew-x: 0;
    --tw-skew-y: 0;
    --tw-scale-x: 1;
    --tw-scale-y: 1;
    --tw-pan-x:  ;
    --tw-pan-y:  ;
    --tw-pinch-zoom:  ;
    --tw-scroll-snap-strictness: proximity;
    --tw-gradient-from-position:  ;
    --tw-gradient-via-position:  ;
    --tw-gradient-to-position:  ;
    --tw-ordinal:  ;
    --tw-slashed-zero:  ;
    --tw-numeric-figure:  ;
    --tw-numeric-spacing:  ;
    --tw-numeric-fraction:  ;
    --tw-ring-inset:  ;
    --tw-ring-offset-width: 0px;
    --tw-ring-offset-color: #fff;
    --tw-ring-color: rgb(59 130 246 / 0.5);
    --tw-ring-offset-shadow: 0 0 #0000;
    --tw-ring-shadow: 0 0 #0000;
    --tw-shadow: 0 0 #0000;
    --tw-shadow-colored: 0 0 #0000;
    --tw-blur:  ;
    --tw-brightness:  ;
    --tw-contrast:  ;
    --tw-grayscale:  ;
    --tw-hue-rotate:  ;
    --tw-invert:  ;
    --tw-saturate:  ;
    --tw-sepia:  ;
    --tw-drop-shadow:  ;
    --tw-backdrop-blur:  ;
    --tw-backdrop-brightness:  ;
    --tw-backdrop-contrast:  ;
    --tw-backdrop-grayscale:  ;
    --tw-backdrop-hue-rotate:  ;
    --tw-backdrop-invert:  ;
    --tw-backdrop-opacity:  ;
    --tw-backdrop-saturate:  ;
    --tw-backdrop-sepia:  
}
.typography {
    line-height: 1;
    --kit-typography-color: inherit;
    --kit-typography-color-inverted: rgb(255, 255, 255);
    --kit-typography-font-size: inherit;
    --kit-typography-leading: 1;
    color: var(--kit-typography-color);
    font-size: var(--kit-typography-font-size);
    font-family: inherit
}
.typography-headline {
    font-family: red;
    --kit-typography-leading: 1em
}"
`;
